### ISSUE TITLE
consult-line: Add consult-line-start-from-current-line customizable variable

### ISSUE THREAD

**Issue by minad** [minad/consult#287]
Generate the candidates such that the current line comes first and then wrap around at the end. Original proposal and discussion on reddit https://www.reddit.com/r/emacs/comments/mxedli/start_search_from_current_line_in_consultline/.

**Comment by oantolin** [comment]
I wanted to do this but have a question about the current code: why is the default candidate calculated the way it is? Why check every single line to see if it is the nearest line to the current one, isn't the current line guaranteed to be among the lines?

**Comment by oantolin** [comment]
Oh, I guess maybe it's done that way because the default candidate is the nearest _non-empty_ line (which is not the current line if the current line is empty).

**Comment by okamsn** [comment]
> Oh, I guess maybe it's done that way because the default candidate is the nearest _non-blank_ line (which is not the current line if the current line is blank).

This is the reasoning, yes.

**Comment by arthurcgusmao** [comment]
Is there a way to show lines that come before the current line before the current line in the minibuffer?

E.g., when I set `consult-line-start-from-top` to `t`, the first line is placed at the top of the minibuffer. I would like it to be placed in its respective order relative to the other lines:
![image](https://user-images.githubusercontent.com/12454729/119261107-1b478700-bbd6-11eb-8cce-c2dff6c1f7a5.png)

**Comment by minad** [comment]
@arthurcgusmao Take a look at the commit message of https://github.com/minad/consult/commit/da542baf6b1216a4ec391f0eef3a8939c43c8ac3.

**Comment by hrehfeld** [comment]
> Or, from another perspective, if I set it to nil, I would like the 'wrapped around' lines to be placed first in the minibuffer:

> Take a look at the commit message of da542ba.

Can you elaborate? Is there a way to toggle the behavior so that lines appear in buffer order, but point starts at the nearest forward match?

**Comment by minad** [comment]
> Is there a way to toggle the behavior so that lines appear in buffer order, but point starts at the nearest forward match?

There is no such way. You can either start at the top or at the current line, but the starting line will always come first. The reason for this is that when you filter, the candidates will be considered top to bottom. In order to achieve the behavior you want one would have to add special casing to the UI-level and this is something Consult generally avoids.

However your question got me thinking again about potential improvements. At least the line number in the prefix could be shown a bit differently in order to indicate that your search wrapped around.

~~~~ 
3 (starting point)
4
1*
2*
~~~~

**Comment by hrehfeld** [comment]
Hmmm, that's too bad. Do you see any way to extend emacs functionality here, if the "viewer" part also supports it? something like `completing-read-offset`?

**Comment by minad** [comment]
Of course one could extend the completion UI to support this. But I don't miss this functionality.

**Comment by minad** [comment]
@hrehfeld See also https://github.com/minad/consult/issues/366 and https://github.com/minad/vertico/issues/92. I think the feature should be implemented in the completion UI (e.g. Vertico), if at all - hopefully it can be done without significant complication of the Vertico code.

**Comment by arthurcgusmao** [comment]
I also miss this functionality @hrehfeld.

If it helps, Vertico has a `vertico--goto` function, that takes the INDEX of a candidate as argument. This could be a place to start.

**Comment by dalanicolai** [comment]
I am not sure if I understand the requests by @arthurcgusmao and @hrehfeld correctly. But maybe the following provides a solution. Personally, I prefer the Swiper behavior which, by default, is to search from top and use the 'current line' as the default, without any wrapping (just show lines before above, and lines after below) Maybe there exists a 'default way' to achieve this, but I could not find it.

Anyway, it can be simply achieved, of course, by just setting the 'first forward' candidate as the default.
For that, we can just add after the candidates the `let*` of the `consult-line` function the following:
```lisp
	 (default (if top
		      (nth (1- (seq-position (mapcar (lambda (s)
						   (cdr (get-text-property 0 'consult-location s)))
						 candidates)
						curr-line
						#'>))
			   candidates)
		    (car candidates))))
```
and then use that value for the default:
```
     :default default
```

I am not sure if this is what you mean, but anyway, this is what makes consult-line behave as I prefer for me.

@minad Is this behavior already provided somehow? Otherwise, wouldn't it be a good idea to simply provide 'this default' as a configurable option?

**Comment by minad** [comment]
@dalanicolai Maybe @gusbrs could point you to the currently best solution using vertico-preselect for consult-line. It still needs a vertico patch though.

**Comment by gusbrs** [comment]
Hi All,
I'm really enjoying this one, so I'm happy to put this in a less hidden place (than in a deeply nested Reddit thread, as it currently is).

My current setup for it is the following (I'm using `el-patch` to patch the relevant function, but you can use the equivalent of your choice, including using the branch, as suggested by minad):

```elisp
(with-eval-after-load 'vertico
  ;; Experimenting with the 'preselect-function' branch.
  ;;
  ;; https://old.reddit.com/r/emacs/comments/107ec1c/consultline_starting_at_point/j45zomv/
  ;; "The function takes the list of candidates and should then return the
  ;; index of the preselected candidate. The only thing which needs to be done
  ;; is then to set vertico-preselect for consult-line/outline, or more
  ;; generally the consult-location completion category via
  ;; vertico-multiform-mode. The preselection function which finds the nearest
  ;; line should be trivial to implement."

  ;; Instead of using the branch, take the "gnu-elpa-dev" version and patch
  ;; the relevant function.
  (el-patch-defun vertico--recompute (pt content)
    "Recompute state given PT and CONTENT."
    (pcase-let* ((table minibuffer-completion-table)
                 (pred minibuffer-completion-predicate)
                 (before (substring content 0 pt))
                 (after (substring content pt))
                 ;; bug#47678: `completion-boundaries` fails for `partial-completion`
                 ;; if the cursor is moved between the slashes of "~//".
                 ;; See also marginalia.el which has the same issue.
                 (bounds (or (condition-case nil
                                 (completion-boundaries before table pred after)
                               (t (cons 0 (length after))))))
                 (field (substring content (car bounds) (+ pt (cdr bounds))))
                 ;; `minibuffer-completing-file-name' has been obsoleted by the completion category
                 (completing-file (eq 'file (vertico--metadata-get 'category)))
                 (`(,all . ,hl) (vertico--all-completions content table pred pt vertico--metadata))
                 (base (or (when-let (z (last all)) (prog1 (cdr z) (setcdr z nil))) 0))
                 (vertico--base (substring content 0 base))
                 (def (or (car-safe minibuffer-default) minibuffer-default))
                 (groups) (def-missing) (lock))
      ;; Filter the ignored file extensions. We cannot use modified predicate for this filtering,
      ;; since this breaks the special casing in the `completion-file-name-table' for `file-exists-p'
      ;; and `file-directory-p'.
      (when completing-file (setq all (completion-pcm--filename-try-filter all)))
      ;; Sort using the `display-sort-function' or the Vertico sort functions
      (setq all (delete-consecutive-dups (funcall (or (vertico--sort-function) #'identity) all)))
      ;; Move special candidates: "field" appears at the top, before "field/", before default value
      (when (stringp def)
        (setq all (vertico--move-to-front def all)))
      (when (and completing-file (not (string-suffix-p "/" field)))
        (setq all (vertico--move-to-front (concat field "/") all)))
      (setq all (vertico--move-to-front field all))
      (when-let (group-fun (and all (vertico--metadata-get 'group-function)))
        (setq groups (vertico--group-by group-fun all) all (car groups)))
      (setq def-missing (and def (equal content "") (not (member def all)))
            lock (and vertico--lock-candidate ;; Locked position of old candidate.
                      (if (< vertico--index 0) -1
                        (seq-position all (nth vertico--index vertico--candidates)))))
      `((vertico--base . ,vertico--base)
        (vertico--metadata . ,vertico--metadata)
        (vertico--candidates . ,all)
        (vertico--total . ,(length all))
        (vertico--highlight . ,hl)
        (vertico--allow-prompt . ,(or def-missing (eq vertico-preselect 'prompt)
                                      (memq minibuffer--require-match
                                            '(nil confirm confirm-after-completion))))
        (vertico--lock-candidate . ,lock)
        (vertico--groups . ,(cadr groups))
        (vertico--all-groups . ,(or (caddr groups) vertico--all-groups))
        (vertico--index . ,(or lock
                               (el-patch-swap
                                 (if (or def-missing (eq vertico-preselect 'prompt) (not all)
                                         (and completing-file (eq vertico-preselect 'directory)
                                              (= (length vertico--base) (length content))
                                              (test-completion content table pred)))
                                     -1 0)
                                 (if (memq vertico-preselect '(prompt directory first))
                                     (if (or def-missing (eq vertico-preselect 'prompt) (not all)
                                             (and completing-file (eq vertico-preselect 'directory)
                                                  (= (length vertico--base) (length content))
                                                  (test-completion content table pred)))
                                         -1 0)
                                   (funcall vertico-preselect all))))))))

  (defun gb/consult-location-preselect (line cands)
    (or (when (= 0 (length cands)) -1)
        (seq-position cands line
                      (lambda (c ol)
                        (>= (cdr (consult--get-location c)) ol)))
        (1- (length cands))))

  (defun gb/consult-location-setup (_)
    (let ((line (with-minibuffer-selected-window
                  (save-excursion
                    (when (eq this-command 'consult-outline)
                      (ignore-errors
                        (outline-back-to-heading t)))
                    (line-number-at-pos (point) consult-line-numbers-widen)))))
      (setq-local vertico-preselect
                  (apply-partially #'gb/consult-location-preselect line))))

  (setq consult-line-start-from-top t)

  ;; The 'consult-location' category includes: 'consult-outline',
  ;; 'consult-line', 'consult-line-multi', 'consult-mark', and
  ;; 'consult-global-mark'.  This preselection setup works well for
  ;; 'consult-outline' and 'consult-line'.  For 'consult-mark' and
  ;; 'consult-global-mark', it makes no sense, since the candidates are not
  ;; sorted by their order in the buffer, but in the order of the mark ring.
  ;; For 'consult-line-multi', the setup is also not appropriate.  It
  ;; appears to work as long as there are candidates for the input given in
  ;; the current buffer, if there are not, the preselection will select a
  ;; candidate of another buffer according to point position on the current
  ;; buffer, which makes no sense.
  (add-to-list 'vertico-multiform-commands
               '(consult-line gb/consult-location-setup))
  (add-to-list 'vertico-multiform-commands
               '(consult-outline gb/consult-location-setup))

  (defun gb/consult-org-heading-preselect (line cands)
    (or (when (= 0 (length cands)) -1)
        (seq-position cands line
                      (lambda (c ol)
                        (>= (thread-last
                              (get-text-property 0 'consult--candidate c)
                              line-number-at-pos
                              with-minibuffer-selected-window)
                            ol)))
        (1- (length cands))))

  (defun gb/consult-org-heading-setup (_)
    (let ((line (with-minibuffer-selected-window
                  (save-excursion
                    (ignore-errors
                      (org-back-to-heading t))
                    (line-number-at-pos (point))))))
      (setq-local vertico-preselect
                  (apply-partially #'gb/consult-org-heading-preselect line))))

  (add-to-list 'vertico-multiform-commands
               '(consult-org-heading gb/consult-org-heading-setup)))
```

@minad This has been through a little polishing since we last discussed this, but it's been stable here for weeks now. As far as my use cases go, and as far as I can see, this is looking good now. I eventually spent some more time thinking about the applicability of this to other commands in the `consult-location` category, and my thoughts are in the comments.

**Comment by geza-herman** [comment]
@minad Wouldn't vertico's design be clearer with allowing `vertico-preselect` to be a function? I mean, `vertico--recompute` currently has a special case for files, but shouldn't it be completely general? As far as I undestand, if `vertico-preselect` could be a function, this special case for files could be moved out of `vertico--recompute`.

And it also would have the secret benefit that people wouldn't need to copy the `vertico-recompute` version mentioned here to their `init.el` file to customize `consult-line`'s initial behavior.

(Thanks for your work! Your packages are awesome. I recently switched from helm. I managed to configure everything to my liking, and it wasn't too hard. The only exception is `consult-line`, which needs this large piece of code to behave as I like)
